Library/Homebrew/formula.rb

```diff
@@ -980,25 +998,45 @@ def migration_needed?
   end
 
   # @private
-  def outdated_versions
-    @outdated_versions ||= begin
-      all_versions = []
-
+  def outdated_versions(options = {})
+    begin
       raise Migrator::MigrationNeededError.new(self) if migration_needed?
 
-      installed_kegs.each do |keg|
-        version = keg.version
-        all_versions << version
-        return [] if pkg_version <= version
+      if @outdated_versions && !options[:fetch_head]
+        return @outdated_versions
+      elsif @outdated_versions_head_fetched && options[:fetch_head]
+        return @outdated_versions_head_fetched
       end
 
+      if options[:fetch_head]
+        @outdated_versions_head_fetched ||= _outdated_versions(options)
+      else
+        @outdated_versions ||= _outdated_versions(options)
+      end
+    end
+  end
+
+  def _outdated_versions(options = {})
+    all_versions = []
+
+    installed_kegs.each do |keg|
+      version = keg.version
+      all_versions << version
+
+      return [] if pkg_version <= version && !version.head?
+    end
+
+    head_version = latest_head_version
+    if head_version
+      head_version_outdated?(head_version, options) ? all_versions.sort! : []
+    else
       all_versions.sort!
```

Comment:
A nitpick: I would prefer to using `all_version.sort` in this line and the line above. After all, you are using the return value.